Add search/filter feature to the "Manage Sensor" screen

chriss158 opened the pull request:
<!-- Thank you for submitting a Pull Request and helping to improve Home Assistant. Please complete the following sections to help the processing and review of your changes. Please do not delete anything from this template. -->

## Summary
<!-- Provide a brief summary of the changes you have made and most importantly what they aim to achieve -->
This PR adds the possibility to hide disabled sensors.

## Screenshots
<!-- If this is a user-facing change not in the frontend, please include screenshots in light and dark mode. -->
<img src="https://user-images.githubusercontent.com/12832850/110859594-f1abde80-82bb-11eb-9052-c918799f49a5.png" width="40%">
<img src="https://user-images.githubusercontent.com/12832850/110859596-f2dd0b80-82bb-11eb-9f9a-47f948184839.png" width="40%">


## Link to pull request in Documentation repository
<!-- Pull requests that add, change or remove functionality must have a corresponding pull request in the Companion App Documentation repository (https://github.com/home-assistant/companion.home-assistant). Please add the number of this pull request after the "#" -->
Documentation: home-assistant/companion.home-assistant#

## Any other notes
<!-- If there is any other information of note, like if this Pull Request is part of a bigger change, please include it here. -->

Additional changes:
* Made "Disabled" sensor summary translatable
* Show "Enabled" as sensor summary if sensor is enabled but has no state

dshokouhi commented on app/src/main/res/values/strings.xml:
```diff
@@ -153,6 +155,7 @@ to your home internet.</string>
   <string name="enable_remaining_sensors">Enable %1$d Sensors</string>
   <string name="enabled_summary">When enabled values will be sent to Home Assistant</string>
   <string name="enabled_title">Enabled</string>
+  <string name="enabled">Enabled</string>
```
should we just stick to using `enabled_title` since the string is the same?

chriss158 replied:
When i think about that now, i would use only one. If the strings need to be different, than we can still change this.

dshokouhi reviewed (approved):
Looks good to me, tested debug apk with a few sensors.  Very minor comment from me since the string seems to be the same.

JBassett commented on app/src/main/java/io/homeassistant/companion/android/sensors/SensorsSettingsFragment.kt:
```diff
@@ -96,6 +102,23 @@ class SensorsSettingsFragment : PreferenceFragmentCompat() {
 
+        findPreference<SwitchPreference>("show_only_enabled_sensors")?.let {
+
+            it.summary = getShowOnlyEnabledSensorsSummary(getShowOnlyEnabledSensors())
```
This can be done in the XML via this: https://developer.android.com/reference/androidx/preference/SwitchPreference#setSwitchTextOff(int)

chriss158 replied:
Thanks i used:  `android:summaryOff` /  `android:summaryOn`

JBassett commented on app/src/main/java/io/homeassistant/companion/android/sensors/SensorsSettingsFragment.kt:
```diff
@@ -184,10 +207,75 @@ class SensorsSettingsFragment : PreferenceFragmentCompat() {
+        if (showOnlyEnabledSensors) showHideGroupsIfNeeded()
+    }
+
+    private fun getShowOnlyEnabledSensors(showOnlyEnabledSensors: Boolean) {
+        runBlocking {
+            integrationUseCase.setShowOnlyEnabledSensors(showOnlyEnabledSensors)
+        }
+    }
+    private fun getShowOnlyEnabledSensors(): Boolean {
+        var showOnlyEnabledSensors = false
+        runBlocking {
+            showOnlyEnabledSensors = integrationUseCase.getShowOnlyEnabledSensors()
+        }
+        return showOnlyEnabledSensors
+    }
```
I don't think we should store this as a setting.  This is just really a filter and should default to off but not be stored.

chriss158 replied:
I thought of this more like a setting rater than a temp. filter. My use case is, that i configure my sensors only one time (after app installation) and then work (check sensors states/reconfigure enabled sensors) from time to time with the sensors. In this use case it is really convenient to just see the enabled sensors, every time i enter the manage sensor screen.

Nevertheless, I like the temp. filter idea. But then i would not use a `SwitchPreference` to filter things. 
I imagine that in the upper right corner there is a filter symbol and a search symbol to find certain sensors.
But to realize this, i need to rework this PR. I think this is the better approach.

dshokouhi replied:
I was under the impression that preferences are always stored and remember what the user selected/typed.  When I check notification history and change the filter the next time I come back it remembers it.  Even after I force stop the app and reopen.

chriss158 replied:
That's interesting. I tried it, without storing a setting, and the setting was reset every time i entered the manage sensor screen. I need to have a look at this again.

chriss158 replied:
Ok i reworked the PR a bit. (See post below) But i still think, its better to save the filters. I think some users would complain about the reset filter. For me it feels not right, when the filter gets reset every time.

@dshokouhi You were right. The preference save it's value automatically. But, now that i use the action bar menu items, i need to use the storage.

nick2525 commented on app/src/main/java/io/homeassistant/companion/android/sensors/SensorsSettingsFragment.kt:
```diff
@@ -184,10 +207,75 @@ class SensorsSettingsFragment : PreferenceFragmentCompat() {
+    private fun showHideGroupsIfNeeded() {
+        for (pref in preferenceScreen) {
+            val prefGroup = pref as PreferenceGroup
+            if (prefGroup != null) {
```
As you cast as "pref as PreferenceGroup" it can't be null, It will be exception if pref is not PreferenceGroup, if you want nullable, cast as "pref as? PreferenceGroup"

chriss158 replied:
Thanks 👍 Changed that

chriss158 commented:
I reworked the PR. Mainly replaced the preference "Show enabled sensors only" by an actionbar item and added a search functionality in the actionbar.

<img src="https://user-images.githubusercontent.com/12832850/112737313-e2828d00-8f59-11eb-8c2b-3f4c235416e5.png" width="40%">
<img src="https://user-images.githubusercontent.com/12832850/112737314-e31b2380-8f59-11eb-9ece-d9296826f702.png" width="40%">
<img src="https://user-images.githubusercontent.com/12832850/112737315-e3b3ba00-8f59-11eb-8b0e-1f76d2cd11ac.png" width="40%">
<img src="https://user-images.githubusercontent.com/12832850/112737316-e44c5080-8f59-11eb-880c-f28560a2586d.png" width="40%">

dshokouhi commented:
I think the filter and search bar make it look clean :) great idea there!

JBassett reviewed (changes requested):
Still not sold on storing the value but we can beta it.
Need to fix merge conflicts then good to merge.

chriss158 commented:
No problem. Let's remove it and see if there are popping up issues. If yes, then we can still implement a setting 👍

chriss158 commented:
I think i misunderstood you. 
Storing a setting is only needed when the setting needs to be restored after the app was force closed. Right? That is not what i want.
I wanted that the setting is hold as long as the app runs. If the app was force closed or killed by the system, it's okay that the setting will be reset.

The `showOnlyEnabledSensors` is now a static class variable. It's value is hold as long the app is in the memory. Am i right here? Till now i thought the value is only hold while the `Manage Sensors` screen is open and will be reset after leaving the screen. But that is  not the case 👍

JBassett reviewed (approved):
This was exactly what I was thinking!
